### ISSUE TITLE
[propose/discuss] Add Pants build tool to coursier

### ISSUE THREAD

**Issue by wisechengyi** [coursier/coursier#695]
A few reasons why SBT isn't ideal: http://www.lihaoyi.com/post/SowhatswrongwithSBT.html

Personally speaking, I feel the build time is getting quite slow, so Pants is definitely going to help with more modularization and build cache. Scala is first class citizen in Pants https://www.pantsbuild.org/

Also an option to use something like bazel, but I am not familiar with that per se.

Curious what is your take on it @alexarchambault

**Comment by wisechengyi** [comment]
I put up a sample at https://github.com/coursier/coursier/pull/697, since proguarded jar behaviors really werid and I couldn't find a way to build non-proguarded one. Sorry a bit messy with other changes, but the part matters is just adding some BUILD files and Pants bootstrap script.
```
$ git diff 659 --stat
 3rdparty/jvm/BUILD                                   |  96 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 cache/src/main/scala/BUILD                           |   9 +++++++++
 cli/src/main/scala-2.11/BUILD                        |  19 +++++++++++++++++++
 cli/src/main/scala-2.11/coursier/cli/Bootstrap.scala |  33 ++-------------------------------
 cli/src/main/scala-2.11/coursier/cli/Coursier.scala  |   2 +-
 core/BUILD                                           |  13 +++++++++++++
 extra/src/main/scala/coursier/extra/BUILD            |   7 +++++++
 pants                                                | 102 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 pants.ini                                            |  13 +++++++++++++
 paths/src/main/java/BUILD                            |   7 +++++++
 10 files changed, 269 insertions(+), 32 deletions(-)
```

You can do something like below:
```
# Run the binary target
$ ./pants run cli/src/main/scala-2.11:coursier-cli -- fetch --help

# Make a distributable bundle at dist/coursier-cli.jar
$ ./pants binary cli/src/main/scala-2.11:coursier-cli
```

The initial bootstrap may take a few seconds, but later runs are much faster than sbt.

**Comment by godenji** [comment]
@wisechengyi would be interesting to compare to latest sbt (1.0.4 released today), which fixes a long standing issue with slow hashing leading to laggy incremental builds .

Ironically enough running `sbt -Dsbt.task.timings=true` now shows `coursierResolutions` and `coursierArtifacts` (triggered by `update` task) as the top consumers. Would *love* to have these tasks fully cached, the incurred overhead really adds up over lifetime of an application.

**Comment by wisechengyi** [comment]
Indeed. I can probably take some timing given the chance. Also it does not hurt to have both build tools in the repo. (Going to change the title)

A couple of wins with Pants:
* Pants is battle tested at large scale repo, so it is mature including the caching bit. 
* Simplicity of build definitions. No crazy/verbose `build.sbt` or `Deps.scala`
* Clear dependency graph. With Pants command, it is simple to see the source/3rdparty dependency graph. In fact, while working on the minimal build for coursier-cli, I discovered some cycles/awkwardness in the build, and I think I've read somewhere that scalac has bad really performance with cycles [citation needed]

**Comment by alexarchambault** [comment]
@wisechengyi I like the idea of having several build tools. Two remarks though:

- I'm personally not too versed into pants for now. You (or anyone else) will have to chip in from time to time to update the pants definitions.
- I don't like pants file lying all around in the sources… Compared to it, sbt only adds `*.sbt` files at the root, plus things under `project/`, and that's it (no sbt specific files in `src` directories). Wouldn't it be possible to keep all the pants things under a specific directory at the root?
- The CI of coursier is currently quite cluttered (usually takes more than 40 minutes to run), so I'm not too keen in running pants from it.
- There's some duplication with the sbt build definitions (dependency versions, …). The pants build definition might not always be up-to-date compared to the sbt ones (all the more if pants isn't run on the CI).

**Comment by wisechengyi** [comment]
Thanks for the comments. @alexarchambault 
 
> I'm personally not too versed into pants for now. You (or anyone else) will have to chip in from time to time to update the pants definitions.

I am fine maintaining the build for the Pants side, as long as I am still around.

> Wouldn't it be possible to keep all the pants things under a specific directory at the root?

That it cannot do unfortunately. Other major build systems such as bazel or buck uses the same scheme.

> The CI of coursier is currently quite cluttered (usually takes more than 40 minutes to run), so I'm not too keen in running pants from it. 
> There's some duplication with the sbt build definitions (dependency versions, …). The pants build definition might not always be up-to-date compared to the sbt ones (all the more if pants isn't run on the CI).

The goal is to have any changes to build and test with sbt as well, so having only sbt and not pants run the tests is ok. However if you happen to like Pants over time, then there is certainly room to migrate over the tests. If it happens, Pants support test sharding natively and can help with the clustering issue. 

There is a bit TODO in that sample which is to get the native portion to build https://github.com/coursier/coursier/pull/697/files#diff-7b18880c85abf3f7e718832370207f7aL42

I can put up a simpler PR with only the build changes without other code change. 
Please let me know if you have more concerns.

**Comment by wisechengyi** [comment]
PR created at https://github.com/coursier/coursier/pull/709

**Comment by jvican** [comment]
I tried the pants build and worked for me, great work @wisechengyi. Would be interesting to know which build runs faster in the CI.

**Comment by wisechengyi** [comment]
@jvican thanks for trying it out! let me know how that goes down the road :)

**Comment by wisechengyi** [comment]
closed with #709
